cpp/CMakeLists.txt

```diff
@@ -904,21 +913,30 @@ set(ARROW_SRCS
   src/arrow/visitor.cc
 
   src/arrow/io/file.cc
-  src/arrow/io/hdfs.cc
-  src/arrow/io/hdfs-internal.cc
   src/arrow/io/interfaces.cc
   src/arrow/io/memory.cc
 
-  src/arrow/ipc/feather.cc
-  src/arrow/ipc/json.cc
-  src/arrow/ipc/json-internal.cc
-  src/arrow/ipc/metadata.cc
-  src/arrow/ipc/reader.cc
-  src/arrow/ipc/writer.cc
-
   src/arrow/util/bit-util.cc
   src/arrow/util/decimal.cc
 )
+if (ARROW_HDFS)
+  set(ARROW_SRCS ${ARROW_SRCS}
+    src/arrow/io/hdfs.cc
+    src/arrow/io/hdfs-internal.cc
+  )
+endif()
```

Comment:
both of them include the header `hdfs.h` located in `${HADOOP_HOME}`. The `ARROW_HDFS` flag if set to `OFF` will not include this header. These files should not be compiled then.